---
title: "lab5_metagenomics"
author: "Ilay Jacques Paz"
date: "2025-10-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

QUESTION 1
```{r}

lapply(c("dada2","phyloseq","vegan","tidyverse",
         "Biostrings"), library, character.only = T)
# All forward and reverse fastq file names have the exact format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
#Note the pattern
data_dir <-"MiSeq_SOP"
fnFs <- sort(list.files(data_dir, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(data_dir, pattern="_R2_001.fastq", full.names = TRUE))
#basename removes the folder names
sample.names <- stringr::str_split_fixed(basename(fnFs),"_",2)[,1]
plotQualityProfile(fnFs[1:38])
plotQualityProfile(fnRs[1:38])
```

QUESTION 2
```{r}
filtFs <- file.path("filtered", paste0(sample.names, "_F_filt.fastq.gz"))
names(filtFs) <- sample.names
#R2
filtRs <- file.path("filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtRs) <- sample.names
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
                     maxN=0, maxEE=c(2,2), truncQ = 2 ,rm.phix=TRUE,
                     compress=TRUE, multithread = TRUE)

#Some reshuffling of the data, get percent of reads left per sample as well as absolute numbers
out2 <- as.data.frame(out) %>% mutate(prop = (100*(reads.out/reads.in)))

# So far looks good
head(out2)
#Mean percent of seqs left
mean(out2$prop)
#Min percent of seqs left
min(out2$prop)
#But what's the sample with the lowest number of reads?
min(out2$reads.out)
errF <- learnErrors(filtFs, multithread=T)
errR <- learnErrors(filtRs, multithread=T)
plotErrors(errR, nominalQ=TRUE)
```
The results seen between forward and reverse reads are similar but the forward reads follow the red trendlines more closely

QUESTION 3
```{r}
dadaFs <- dada(filtFs, err=errF, multithread = TRUE)
dadaRs <- dada(filtRs, err=errF, multithread = TRUE)
dadaFs[[1]]
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)

# Inspect the merger data.frame from the first sample
DT::datatable(mergers$F3D0,options = list(scrollX = TRUE, autoWidth = TRUE))
seqtab <- makeSequenceTable(mergers)

dim(seqtab)
table(nchar(getSequences(seqtab)))
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=F, verbose=TRUE)
#Proportion ASVs were removed due to chimeras?
100-(dim(seqtab.nochim)[2]/dim(seqtab)[2]*100)
sum(seqtab.nochim)/sum(seqtab)*100
ASV_seqs <- DNAStringSet(base::colnames(seqtab.nochim) %>% 
                           `names<-`(base::paste("ASV",seq(1:ncol(seqtab.nochim)),sep = "")))

taxa <- assignTaxonomy(seqtab.nochim, "silva_nr99_v138.1_wSpecies_train_set.fa.gz", multithread=TRUE)
taxa2 <- as.data.frame(taxa) %>% 
  `row.names<-`(base::paste("ASV", seq(1:nrow(taxa)), sep = ""))
taxa2 <- taxa2 %>% filter(Family != "Mitochondria") %>% 
  filter(Order != "Chloroplast")
# Firmicutes
firmicutes_ids <- rownames(taxa2)[taxa2$Phylum == "Firmicutes"]
firmicutes_seqs <- ASV_seqs[names(ASV_seqs) %in% firmicutes_ids]
genus <- as.character(taxa2[firmicutes_ids, "Genus"])
names(firmicutes_seqs) <- paste(firmicutes_ids, genus)
writeXStringSet(firmicutes_seqs, "Firmicutes_ASVs_identified.fa")
head(firmicutes_seqs)

# Actinobacteria
actino_ids <- rownames(taxa2)[taxa2$Phylum %in% c("Actinobacteria", "Actinobacteriota")]
actino_seqs <- ASV_seqs[names(ASV_seqs) %in% actino_ids]
genus_actino <- as.character(taxa2[actino_ids, "Genus"])
names(actino_seqs) <- paste(actino_ids, genus_actino)
writeXStringSet(actino_seqs, "Actinobacteria_ASVs_identified.fa")
actino_seqs
```
ASV 22 (FIRMICUTES) AND ASV 207 (ACTINOBACTERIA) UNIDENTIFIED WILL BE BLASTED





ASV 207 BLAST results it could be Adlercreutzia, Eggerthella,or Enteroscipio
 ![ASV 207 BLAST results it could be Adlercreutzia, Eggerthella,or 	Enteroscipio ](/Users/ipaz00/Desktop/asv207.png) 
 ASV 22 BLAST results, it could be any bacteria
![ASV 22 BLAST results, it could be any bacteria ](/Users/ipaz00/Desktop/asv22.png) 
